File: frontend/src/types/donation.ts
// Donation related types
export const DonationStatus = {
  Diajukan: "Diajukan",
  SiapDijemput: "Siap Dijemput",
  Diterima: "Diterima",
} as const;

export type DonationStatus =
  (typeof DonationStatus)[keyof typeof DonationStatus];

export interface Donation {
  donation_id: number;
  donor_user_id: number;
  receiver_user_id?: number;
  type_of_food: string[];
  latitude: number;
  longitude: number;
  status: DonationStatus;
  created_at: Date;
  expires_at: Date;
  // Relations
  donor_name?: string;
  receiver_name?: string;
}

export interface DonationCreate {
  donor_user_id: number;
  type_of_food: string[];
  latitude: number;
  longitude: number;
  status: DonationStatus;
}

export interface DonationUpdate {
  type_of_food?: string[];
  latitude?: number;
  longitude?: number;
  status?: DonationStatus;
  receiver_user_id?: number;
}

export interface DonationSearch {
  latitude: number;
  longitude: number;
  radius_km?: number;
  status?: DonationStatus;
}

export interface NearbyDonation extends Donation {
  distance_km: number;
}

File: backend/models/donation.py
"""
Donation model and schema definitions
"""
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime
from enum import Enum

class DonationStatus(str, Enum):
    DIAJUKAN = "Diajukan"
    SIAP_DIJEMPUT = "Siap Dijemput"
    DITERIMA = "Diterima"

class DonationBase(BaseModel):
    donor_user_id: int  # Foreign key to accounts.user_id (who donates)
    type_of_food: List[str]  # List of food types
    latitude: float
    longitude: float
    status: DonationStatus = DonationStatus.DIAJUKAN

class DonationCreate(DonationBase):
    pass

class DonationUpdate(BaseModel):
    type_of_food: Optional[List[str]] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    status: Optional[DonationStatus] = None
    receiver_user_id: Optional[int] = None  # Who receives the donation

class DonationResponse(DonationBase):
    donation_id: int
    receiver_user_id: Optional[int] = None  # Foreign key to accounts.user_id (who receives)
    created_at: datetime
    expires_at: datetime  # 1 hour after created_at
    
    class Config:
        from_attributes = True

class DonationInDB(DonationResponse):
    pass